ARCHITECTURE.md

```diff
@@ -0,0 +1,265 @@
+the call.
+
+#### Login
+
+The login operation is initiated by the Client. It may result in one of the following:
+- a redirection of the Resource Owner to the Issuer's authorization endpoint.
+- a request by the Client to the Issuer's token endpoint.
+
+Handlers for the login operation are located in `packages/*/src/login/oidc/oidcHandlers/*Handler.ts`.
+
+#### Incoming redirect
+
+The incoming redirect operation is initiated by the Issuer.
+At the Issuer webpage, the Resource Owner authenticates (e.g., by entering a username
+and a password), after which the Issuer sends them to a webpage under the Client
+app's control (specifically its `redirect_uri`). The Issuer appends some very important query 
+parameters to the IRI the Resource Owner is redirected to, which are needed by the Client
+to complete the login flow. This is done when the developer calls `handleIncomingRedirect()`,
+and the Handlers for the incoming redirect are located
+in `packages/*/src/login/oidc/redirectHandler/*Handler.ts`.
+
+### Dependency injection
+
+An important architectural component of this library is dependency injection,
+implemented here using [TSyringe](https://github.com/Microsoft/tsyringe). Dependencies
+are declared in `packages/*/src/dependencies.ts`.
+
+#### Declaring order
+
+The order in which the dependencies are declared (for a given container) **matters**.
+Let's have a look at some code to make things clearer.
+
+```
+container.register<IOidcHandler>("browser:oidcHandler", {
+  useClass: AggregateOidcHandler,
+});
+container.register<IOidcHandler>("browser:oidcHandlers", {
+  useClass: RefreshTokenOidcHandler,
+});
+container.register<IOidcHandler>("browser:oidcHandlers", {
+  useClass: AuthorizationCodeWithPkceOidcHandler,
+});
+container.register<IOidcHandler>("browser:oidcHandlers", {
+  useClass: ClientCredentialsOidcHandler,
+});
+```
+
+Here, `AggregateOidcHandler` is the handler aggregator (as defined in the Handler
+Pattern section), and `RefreshTokenOidcHandler`, `AuthorizationCodeWithPkceOidcHandler`
+and `ClientCredentialsOidcHandler` are its underlying handlers. When receiving a
+request, `AggregateOidcHandler` will first invoke its instance of `RefreshTokenOidcHandler`
+to check if it can handle it. If so, that instance of `RefreshTokenOidcHandler`
+will handle the request, and the instances of `AuthorizationCodeWithPkceOidcHandler`
+and `ClientCredentialsOidcHandler` will not be called. This means that it is
+important to declare the dependencies from the most specialized to the most generic,
+because if a fallback handler that can handle all requests is declared first,
+the other more specialized handlers will never be called.
+
+Note that the label for the containers of the aggregator and the underlying handlers
+differ in the example above:
+- `browser:oidcHandler` (without an 's') for the aggregator.
+- `browser:oidcHandlers` (with an 's'), for the aggregated handlers.
```

Comment:
I would probably move this up a little?  

Here, `AggregateOidcHandler` is the handler aggregator (as defined in the Handler
Pattern section), and `RefreshTokenOidcHandler`, `AuthorizationCodeWithPkceOidcHandler`
and `ClientCredentialsOidcHandler` are its underlying handlers. 

Note that the label for the aggregator containers differ from that for the underlying handlers (or something like that):
- `browser:oidcHandler` (without an 's') for the aggregator
- `browser:oidcHandlers` (with an 's') for the aggregated handlers.

When receiving a
request, `AggregateOidcHandler` will first invoke its instance of `RefreshTokenOidcHandler`
to check if it can handle it. If so, that instance of `RefreshTokenOidcHandler`
will handle the request, and the instances of `AuthorizationCodeWithPkceOidcHandler`
and `ClientCredentialsOidcHandler` will not be called. This means that it is
important to declare the dependencies from the most specialized to the most generic,
because if a fallback handler that can handle all requests is declared first,
the other more specialized handlers will never be called.

// then we can just flow into the sections when the order isn't relevant